added newLit(NimNode) overload

timotheecour opened the pull request:
* this PR takes over the 1st half of https://github.com/nim-lang/Nim/pull/13888 which was stale
* please don't squash so that @krux02 gets credited for the 1st commit
* I reverted the change to `astGenRepr` which was the main contentious point from https://github.com/nim-lang/Nim/pull/13888 (the 2nd half)
* I agree with rationale from #13888 for keeping the name `newLit` (among other reasons, newLit already works on seq etc, and it'd be gratuitious for a new name to be required)

## future work
* the improvements from https://github.com/nim-lang/Nim/pull/13888 to `astGenRepr` can be done in a separate PR by adding an additional param for backward compatibility (details TBD)
* https://github.com/nim-lang/Nim/pull/13888 worsened output for string litterals so it should be fixed in the PR that modifies `astGenRepr`
```
-      newCommentStmtNode(
-        "This is a docstring"
-      ),
+    nnkStmtList.newTree(
+      newCommentStmtNode("This is a docstring"),
```
* maybe repr can be improved (in future PR) to avoid the formatting code from https://github.com/nim-lang/Nim/pull/13888

krux02 committed:
implement newLit on NimNode and make astGenRepr to use it

timotheecour committed:
revert changes to astGenRepr; s/newIdentNode/ident; changelog; since

alaviss commented:
@Araq, if you want to squash, add this to the end of the commit description before merging (Github should really start being smart enough to do this automatically):

```
Co-authored-by: Arne Döring <arne.doering@gmx.net>
```

Araq commented:
The name ``newLit`` has nothing to do with what the proc actually does. There are no use cases given for this proc, no examples.  

> some nodes kinds have constructor procs

I don't agree with this special casing. The constructor procs are all redundant and shouldn't be produced at all. There is only ``newLit`` for atoms and ``newTree`` for composed nodes.

timotheecour committed:
rename newLit(NimNode) => astGen

timotheecour committed:
fixup

timotheecour commented:
PTAL
> The name newLit has nothing to do with what the proc actually does.

renamed to `astGen`

> There are no use cases given for this proc, no examples.

fixed, and explained use case in doc comment+runnableExamples

> The constructor procs are all redundant and shouldn't be produced at all

the point is to produce as idiomatic code as possible that one can use when writing macros, contrary to what's produced by `astGenRepr`.
I added a `compact=false` optional param to choose between canonical representation and idiomatic one.

I added a disclaimer regarding that output is subject to change.

timotheecour committed:
add compact param
